Ap 26816 create dockerfile

gianlucabiccari opened the pull request:
PR to merge the files necessary to create a docker image and to run the container. I tried to follow some of the docker best practices (after discussing this with Smith). When this is merged I will use the cookie-cutter to start with the de-identification repo. After that I am sure we will find better way to do things and we will come back to fix the cookie cutter incrementally.

gianlucabiccari committed:
Add dockerignore file

gianlucabiccari committed:
Start writing dockerfile

gianlucabiccari committed:
Add system packages

gianlucabiccari committed:
Add files for docker

gianlucabiccari committed:
Fixed Indentation

gianlucabiccari committed:
Merge from develop

eabruzzese commented on {{cookiecutter.project_dash_slug}}/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+# Base Image
+FROM python:3.6.6-alpine3.7
```
@gianlucabiccari a few notes here:

1. It's best to not overspecify the version number: 3.6 is enough detail.
2. I like alpine, but in the Python ecosystem your dependencies will install _much_ faster if you use a Debian base (because there are prebuilt wheels for Debian). I use `python:3.6-slim-stretch`, which is comparable in size to alpine and gives us that speed advantage.

gianlucabiccari replied:
I am ok moving to `python:3.6-slim-stretch`, in fact I think we should uniform that across teams that build python projects

eabruzzese commented on {{cookiecutter.project_dash_slug}}/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+
+# Run as non root user (best practice)
+ENV USERNAME=syapse
+ENV PROJECT_NAME={{cookiecutter.project_dash_slug}}
```
I had originally done this in the Flask cookiecutter, where I named users and directories and stuff. But then I realized: the only thing running in this container is this project, and there's only one ever one user, so there's no need to give them special names: I just called the user "user" and put the package in `/opt/package`. It simplified the Dockerfile quite a lot and makes it the same everywhere.

eabruzzese commented on {{cookiecutter.project_dash_slug}}/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+WORKDIR /home/$USERNAME/$PROJECT_NAME
+COPY . .
+RUN addgroup -g 303 -S $USERNAME \
+    && adduser -u 303 -S -G $USERNAME -h /home/$USERNAME $USERNAME
```
You can simplify this block a lot: when you create a user, it'll automatically create a group and home directory for it:

```bash
# Creates a user, its home directory, and a group with the same name.
adduser $USERNAME

# Even simpler, if you took my other suggestion:
adduser user
```

gianlucabiccari replied:
@eabruzzese  I don't like the idea of having a user names user ? can we just call it something else ?

kojiromike replied:
@gianlucabiccari Just keep in mind that we don't need many users -- it's just the super user and the not-super user... What would you like to call it?

gianlucabiccari replied:
syapse ?

eabruzzese commented on {{cookiecutter.project_dash_slug}}/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+COPY . .
+RUN addgroup -g 303 -S $USERNAME \
+    && adduser -u 303 -S -G $USERNAME -h /home/$USERNAME $USERNAME
+RUN chown -R $USERNAME:$USERNAME /home/$USERNAME/$PROJECT_NAME
```
In the future, another thing to keep in mind: since Docker caches in layers, it's often smart to combine them. In this case, you could just add it to the above chain of `RUN` directives.

gianlucabiccari replied:
Good point! I will change this

eabruzzese commented on {{cookiecutter.project_dash_slug}}/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+USER $USERNAME
+
+# Install or update system dependencies
+ENV PIP_NO_CACHE_DIR=off
```
Here are a few other helpful environment variables for Python and Pipenv:

```Dockerfile
##
# Some helpful environment variables:
#
#   - PYTHONFAULTHANDLER enables the fault handler: 
#       -> https://docs.python.org/3/library/faulthandler.html
#   - PYTHONUNBUFFERED ensures that you're getting realtime output when you tail the logs in the container: 
#       -> https://docs.python.org/3/using/cmdline.html#cmdoption-u
#   - PIP_DISABLE_PIP_VERSION_CHECK suppresses the warning message about outdated pip (useful since you're upgrading it on build here).
#   - PIPENV_COLORBLIND disables colored output in pipenv (it plays nicer with CI)
#   - PIPENV_NOSPIN disables the spinner (again, plays nicer with simpler terminals)
ENV PYTHONFAULTHANDLER=1 \
    PYTHONUNBUFFERED=1 \
    PIP_DISABLE_PIP_VERSION_CHECK=on \
    PIPENV_COLORBLIND=true \
    PIPENV_NOSPIN=true
```

eabruzzese commented on {{cookiecutter.project_dash_slug}}/docker-entrypoint.sh:
```diff
@@ -0,0 +1,17 @@
+}
+
+pipenv_shell() {
+  pipenv shell
```
Question: what's the goal of the Docker image? If it's not just for development, we should make the default command something like `pipenv run "$@"`.

gianlucabiccari replied:
The goal is to run the python module developed and exit. In the case of de-identification for example something like `python deidentify --customer miami`

gianlucabiccari replied:
Also this script I was kind of flying blind in the effort to make it generic. your suggestion is good.

kojiromike commented on {{cookiecutter.project_dash_slug}}/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+    python3-dev
+
+# Run as non root user (best practice)
+ENV USERNAME=syapse
```
Is there any value in being flexible here?

gianlucabiccari replied:
no really. assuming that once I picked one username then I will *never* change it. avoiding to change  _n_ lines manually.

kojiromike commented on {{cookiecutter.project_dash_slug}}/docker-entrypoint.sh:
```diff
@@ -0,0 +1,17 @@
+  [ $# -lt 1 ] && pipenv_shell
+  case "$1" in
+    -*) pipenv_shell "$@";;
+    *) exec "$@";;
```
You might want to do `exec` inside the `pipenv_shell` function for consistency, so that all of the code paths effectively end this script.

gianlucabiccari committed:
Add dev image build

gianlucabiccari committed:
minor fix to prevent wrong expansion of command

gianlucabiccari committed:
Use bash to run scripts

gianlucabiccari committed:
Default command run in the virtualenv

gianlucabiccari committed:
Add config part

gianlucabiccari committed:
entry point

gianlucabiccari committed:
More configuration for the docker image

gianlucabiccari committed:
Remove wrong print

gianlucabiccari commented:
@eabruzzese @kojiromike can you review it one more time and approve it if you feel comfortable with that.

eabruzzese committed:
Simplify the Dockerfile, entrypoint

The `Dockerfile` was simplified to keep the image minimal and fast to
build. Specifically:

* Non-critical system package installation was removed.
* The Pipfile, setup.cfg, and setup.py files are copied and installed first.
* The rest of the code much closer to the end of the cli image build.
* The `Pipfile.lock` is tested for and respected.
* The running user was renamed to `user` to keep the package generic.

eabruzzese committed:
Update the docker-entrypoint

The `docker-entrypoint.sh` was updated to include newrelic integration.
This required the package to be added to the `setup.cfg`.

eabruzzese committed:
Use project_dash_slug for the image name.

Docker image names cannot contain spaces, so we use the
`project_dash_slug` instead of the `project_name` to prevent errors.

eabruzzese committed:
Update packaging strategy.

Package-critical dependencies (non-dev dependencies) were put into the
`Pipfile` instead of the `setup.cfg` because they're environmental, and
non critical to the operation of the package.

eabruzzese committed:
Update CLI-related code removal

The `post_gen_project.sh` hook was updated to remove the `Dockerfile`,
`docker-entrypoint.sh`, `bin/run`, and `bin/build` files. If there's no
CLI, there's no need for a Docker image.

eabruzzese committed:
Update bin scripts

In development, the image will need the `.egg-info` directory in order
to run the CLI. The `bin/run` script was updated to include that step
before running the tool. The `bin/build` script was updated for message
consistency.

eabruzzese committed:
Remove extra settings

Some of the settings brought over from the Flask cookiecutter were
specific to some of the packages that were used in it. Those settings were
removed.
